### PR TITLE
have activelaneid_u32 call ockl_activelane_u32

### PR DESCRIPTION
**Pull request by david-salinas**
for https://github.com/RadeonOpenCompute/hcc/issues/1011

### CODE REVIEW

**Comment by b-sumner** [comment]
Is there a test that checks that all/most of the __foo functions compile?  If so it needs updates too.

**Comment by david-salinas** [comment]
> Is there a test that checks that all/most of the __foo functions compile? If so it needs updates too.

sorry, I should have already had a test ready.  Let me cook one up, and update this PR.

File: tests/Unit/AMDGPU/activelaneid.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        test[i * WAVEFRONT_SIZE + j] = 0;
+      }
+    }
+
+    // randomly shuffle items in the block
+    for (int j = 0; j < WAVEFRONT_SIZE * 10; ++j) {
+      int k1 = int_dist(rd);
+      int k2 = int_dist(rd); 
+      if (k1 != k2) {
+        test[i * WAVEFRONT_SIZE + k1] ^= test[i * WAVEFRONT_SIZE + k2] ^= test[i * WAVEFRONT_SIZE + k1] ^= test[i * WAVEFRONT_SIZE + k2];
```
Comment (b-sumner) [inline tests/Unit/AMDGPU/activelaneid.cpp]: Why not just code a swap?

File: tests/Unit/AMDGPU/activelaneid.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      }
+    }
+  }
+
+  for (int i = 0; i < WAVEFRONT_SIZE; ++i) {
+    for (int j = 0; j < WAVEFRONT_SIZE; ++j) {
+#if TEST_DEBUG
```
Comment (b-sumner) [inline tests/Unit/AMDGPU/activelaneid.cpp]: I would move this outside the entire loop nest

File: tests/Unit/AMDGPU/activelaneid.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+#endif
+  }
+
+  array<uint32_t, 1> test_GPU(GRID_SIZE);
+  copy(test.begin(), test_GPU);
+
+  array<uint32_t, 1> output_GPU(GRID_SIZE);
+  extent<1> ex(GRID_SIZE);
+  parallel_for_each(ex, [&](index<1>& idx) [[hc]] {
+    output_GPU(idx) = __activelaneid_u32();
```
Comment (b-sumner) [inline tests/Unit/AMDGPU/activelaneid.cpp]: Every lane is sure to be active.  Control flow is needed to deactivate some lanes.

File: tests/Unit/AMDGPU/activelaneid.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      std::cout << test[i * WAVEFRONT_SIZE + j] << " ";
+#endif
+    }
+#if TEST_DEBUG
+    std::cout << "\n";
+#endif
+  }
+
+  array<uint32_t, 1> test_GPU(GRID_SIZE);
+  copy(test.begin(), test_GPU);
```
Comment (scchan) [inline tests/Unit/AMDGPU/activelaneid.cpp:63]: `test_GPU` is not being used so the data generation in `test` is redundant so they should be removed

File: include/hc.hpp [hunk 5]
```diff
@@ -2817,7 +2817,11 @@ extern "C" inline __attribute((always_inline)) std::uint64_t __cycle_u64() __HC_
  *
  * @return The result will be in the range 0 to WAVESIZE - 1.
  */
-extern "C" unsigned int __activelaneid_u32() __HC__;
+extern "C" __attribute__((always_inline))
```
Comment (scchan) [inline include/hc.hpp]: I don't think this needs the always_inline attribute here

File: include/hc.hpp [hunk 6]
```diff
@@ -2817,7 +2817,11 @@ extern "C" inline __attribute((always_inline)) std::uint64_t __cycle_u64() __HC_
...
+unsigned int __ockl_activelane_u32(void);
+extern "C" unsigned int __activelaneid_u32() __HC__ {
```
Comment (scchan) [inline include/hc.hpp]: this would be marked as `inline` to avoid multiple definitions

**Review by scchan — CHANGES_REQUESTED** [review]

**Comment by scchan** [comment]
@david-salinas  ping...

File: tests/Unit/AMDGPU/activelaneid.cpp (hunk 1 above)
Comment (david-salinas) [reply tests/Unit/AMDGPU/activelaneid.cpp]: Looking at this test (I based it on the existing "activelanecount.cpp" test) most of this set up of "test" can be scrapped, I think.  We can just create a PFE, with a suitable extent, and an output grid/vector and call activelaneid to verify that we get correct/valid laneid values.

File: tests/Unit/AMDGPU/activelaneid.cpp (hunk 4 above)
Comment (david-salinas) [reply tests/Unit/AMDGPU/activelaneid.cpp:63]: ok.  I'm thinking I can scrap "test" too.  I based this test on "activelanecount.cpp".  For activelaneid, we really just need a PFE and an output grid, to verify activelaneid is returning the correct lane id.  I think I can reduce this test, by a lot.

**Comment by b-sumner** [comment]
I'm not sure I'm seeing the latest version of the test code.  I agree it can be small, but the kernel should have some control flow so that we can see that activelaneid is reporting the correct ID on lanes that are truly active.

**Comment by scchan** [comment]
I don't see any update to the code either

**Comment by david-salinas** [comment]
Sorry,  I was about to push an update test, when I realized I need to add some control flow to deactivate some lanes.  I should have an update shortly.

**Review by scchan — APPROVED** [review]

**Pull request merged**